src/rand.c

```diff
@@ -0,0 +1,231 @@
+{
+	HCRYPTPROV provider;
+	SYSTEMTIME systemtime;
+	FILETIME filetime, idletime, kerneltime, usertime;
+
+	if (CryptAcquireContext(&provider, 0, 0, PROV_RSA_FULL,
+	                        CRYPT_VERIFYCONTEXT|CRYPT_SILENT)) {
+		BOOL success = CryptGenRandom(provider, sizeof(uint64_t), (void *)seed);
+		CryptReleaseContext(provider, 0);
+
+		if (success)
+			return 0;
+	}
+
+	GetSystemTime(&systemtime);
+	if (!SystemTimeToFileTime(&systemtime, &filetime)) {
+		git_error_set(GIT_ERROR_OS, "could not get time for random seed");
+		return -1;
+	}
+
+	/* Fall-through: generate a seed from the time and system state */
+	*seed = 0;
+	*seed |= ((uint64_t)filetime.dwLowDateTime << 32);
+	*seed |= ((uint64_t)filetime.dwHighDateTime);
+
+	*seed ^= ((uint64_t)GetCurrentProcessId() << 32);
+	*seed ^= ((uint64_t)GetCurrentThreadId());
+
+	GetSystemTimes(&idletime, &kerneltime, &usertime);
+	*seed |= ((uint64_t)idletime.dwLowDateTime);
+	*seed |= ((uint64_t)idletime.dwHighDateTime << 32);
+	*seed |= ((uint64_t)kerneltime.dwLowDateTime << 32);
+	*seed |= ((uint64_t)kerneltime.dwHighDateTime);
+	*seed |= ((uint64_t)usertime.dwLowDateTime);
+	*seed |= ((uint64_t)usertime.dwHighDateTime << 32);
+
+	/* Mix in the addresses of some functions and variables */
+	*seed ^= (((uint64_t)((uintptr_t)getseed)) << SEED_SHIFT);
+	*seed ^= (((uint64_t)((uintptr_t)seed)));
+	*seed ^= (((uint64_t)((uintptr_t)printf)) << SEED_SHIFT);
+	*seed ^= (((uint64_t)((uintptr_t)&errno)));
+
+	*seed ^= ((uint64_t)git__timer());
+
+	return 0;
+}
+
+#else
+
+GIT_INLINE(int) getseed(uint64_t *seed)
+{
+	struct timeval tv;
+	double loadavg[3];
+	int fd;
+
+# if defined(GIT_RAND_GETENTROPY)
+	GIT_UNUSED((fd = 0));
+
+	if (getentropy(seed, sizeof(uint64_t)) == 0)
+		return 0;
+# else
+	/*
+	 * Try to read from /dev/urandom; most modern systems will have
+	 * this, but we may be chrooted, etc, so it's not a fatal error
+	 */
+	if ((fd = open("/dev/urandom", O_RDONLY)) >= 0) {
+		ssize_t ret = read(fd, seed, sizeof(uint64_t));
+		close(fd);
+
+		if (ret == (ssize_t)sizeof(uint64_t))
+			return 0;
+	}
+# endif
+
+	/* Fall-through: generate a seed from the time and system state */
+	*seed = 0;
+
+	if (gettimeofday(&tv, NULL) < 0) {
+		git_error_set(GIT_ERROR_OS, "could get time for random seed");
+		return -1;
+	}
+
+	*seed = 0;
+	*seed |= ((uint64_t)tv.tv_usec << 40);
+	*seed |= ((uint64_t)tv.tv_sec);
+
+	*seed ^= ((uint64_t)getpid() << 32);
+	*seed ^= ((uint64_t)getpgid(0));
+	*seed ^= ((uint64_t)getppid() << 32);
+	*seed ^= ((uint64_t)getsid(0));
+	*seed ^= ((uint64_t)getuid() << 32);
+	*seed ^= ((uint64_t)getgid());
+
+	*seed ^= ((uint64_t)loadavg[0]);
+	*seed ^= ((uint64_t)loadavg[1]);
+	*seed ^= ((uint64_t)loadavg[2]);
+
+	/* Mix in the addresses of some functions and variables */
+	*seed ^= (((uint64_t)((uintptr_t)getseed)) << SEED_SHIFT);
+	*seed ^= (((uint64_t)((uintptr_t)seed)));
+	*seed ^= (((uint64_t)((uintptr_t)printf)) << SEED_SHIFT);
+	*seed ^= (((uint64_t)((uintptr_t)errno)));
```

Comment:
I'm thinking that `getseed` and `printf` are the same for all threads here and probably doesn't add much. (And because of some microcontrollers with separate address spaces for functions and data C99 doesn't allow casting function pointers to data pointers.)

Are `seed` and `errno` at unique addresses for each thread?
Otherwise `&tv` should give a pointer to a stack variable.

Reply:
I wasn't aiming for uniqueness across threads - we'll seed once for the library at `git_libgit2_init` time for all threads - I was aiming for uniqueness across process invocations to avoid a replay attack.  But for `pthread`, that will be a noop on anything that doesn't do ASLR, which I actually assumed was more common than it is - my Mac doesn't do ASLR, at least not by default.  🤔

And that should have been `&errno`, not `errno`.  Meh.

In any case, I think that I got a bit belt-and-suspenders here; certainly I can remove the function pointers here for broader compatibility, especially since it's not the macOS and Linux cases that are going to fall into this function.

Reply:
Sorry, I got a bit of a tunnel vision here. The cases that doesn't have either `getentropy` or `/dev/urandom` probably lacks most other security functions.

Reply:
I do, too, at this point.  This feels like a lot of code for "temp file in a place that we own", but there are always surprises.